proposals/20230530-driver-kernel-testing-framework.md

```diff
@@ -0,0 +1,170 @@
+
+
+#### Test Infrastructure
+
+Our goal is to facilitate the expansion and integration of full Continuous Integration (CI) for the kernel test grid, enabling comprehensive testing of the kernel driver functionality. Concurrently, we will develop a test framework utilizing localhost virtual machines (VMs), which will include a limited kernel test grid. This framework will be accessible to adopters, providing them with a convenient option for local testing during the development or testing phases.
+
+
+#### Current State
+
+Currently (as of May 31, 2023), the relevant [falcosecurity/libs](https://github.com/falcosecurity/libs) CI drivers tests include:
+
+- Building kernel drivers for:
+    - Latest Archlinux kernel to spot possible incompatibilities with the latest kernel tree changes
+    - ubuntu-22.04 (x86_64)
+    - linux-2.6.32
+    - linux-3.10
+    - linux-4.18
+    - linux-5.19
+    - linux-6.2
+- Running [drivers_test](https://github.com/falcosecurity/libs/tree/master/test/drivers) unit tests on:
+    - ubuntu-22.04 (x86_64)
+    - ubuntu-2204:2022.10.2 (arm64)
+    - ubuntu-2004:202107-02 (x86_64)
+
+
+#### Desired `kernel test grid` Expansion
+
+*Distributions*
+
+Choose a minimum of five popular distributions from the pool of distributions for which Falco currently publishes kernel drivers (retrieved from [falcosecurity/kernel-crawler](https://github.com/falcosecurity/kernel-crawler/tree/kernels) on May 31, 2023). Ensure a balanced representation between deb-based and rpm-based distributions.
+
+- AliyunLinux
+- AlmaLinux
+- AmazonLinux
+- AmazonLinux2
+- AmazonLinux2022
+- AmazonLinux2023
+- ArchLinux
+- BottleRocket
+- CentOS
+- Debian
+- Fedora
+- Flatcar
+- Minikube
+- OpenSUSE
+- OracleLinux
+- PhotonOS
+- RockyLinux
+- Talos
+- Ubuntu
+
+*Kernel versions*
+
+To achieve comprehensive coverage, the statistical sampling of versions across distributions will prioritize the minimum and maximum supported kernel versions per driver type. Additionally, a particular emphasis will be placed on selecting LTS (Long-Term Support) releases when choosing the remaining kernel versions. As part of this approach, a minimum of 10 kernel versions will be carefully selected for each driver type.
+
+*Architectures*
+
+Place higher priority on testing for `x86_64` compared to `aarch64`.
```

Comment:
Fair point, can tweak.

Perhaps another good discussion point: `s390x ` is supported by libs but not officially by Falco. Maybe I can say that because of this initially we prioritize x86 and aarch64 tests, subject to change in future iterations of the test framework ...

@hbrueckner would you have ideas on how we can test `s390x ` specifically for libs better? And ideally cut down the durations of s390x libs builds as well, they take a fairly long time atm. Thank you!

Reply:
My recommendation is:

- Falco officially supported architectures should be prioritized by their adoption level (ie. `x86_64` is P0, `aarch64` is P1)
- just best effort (low priority) for any other arch supported by libs (ie. `s390x` is Nice-To-Have).

Reply:
Love it, this way we clearly articulate priorities.